src/impl_/concat.rs

```diff
@@ -6,25 +6,210 @@ macro_rules! const_concat {
         $e
     }};
     ($l:expr, $($r:expr),+ $(,)?) => {{
-        const L: &'static str = $l;
-        const R: &'static str = $crate::impl_::concat::const_concat!($($r),*);
-        const LEN: usize = L.len() + R.len();
-        const fn combine(l: &'static [u8], r: &'static [u8]) -> [u8; LEN] {
-            let mut out = [0u8; LEN];
-            let mut i = 0;
-            while i < l.len() {
-                out[i] = l[i];
-                i += 1;
+        const PIECES: &[&str] = &[$l, $($r),+];
+        const RAW_BYTES: &[u8] = &$crate::impl_::concat::combine_to_array::<{
+            $crate::impl_::concat::combined_len(PIECES)
+        }>(PIECES);
+        // Safety: `RAW_BYTES` is combined from valid &str pieces
+        unsafe { ::std::str::from_utf8_unchecked(RAW_BYTES) }
+    }}
+}
+
+pub use const_concat;
+
+/// Calculates the total byte length of all string pieces in the array.
+///
+/// This is a useful utility in order to determine the size needed for the constant
+/// `combine` function.
+pub const fn combined_len(pieces: &[&str]) -> usize {
```

Comment:
Given that `&str` and `&[u8]` are layout compatible, I think we could transmute `&[&str]` to `&[&[u8]]` and call into `combined_len_bytes` to avoid the duplication (and the same for the others I guess). I did a quick test in miri on the playground and it did not complain.

Reply:
Given https://github.com/PyO3/pyo3/pull/5286#discussion_r2264782066 and that the cast feels a bit unfortunate (even if totally sound in this case) I might accept the duplication for now and let it get tidied up in a follow up.

Reply:
Sure, sounds good.